server/comms/wsclient.go

```diff
@@ -0,0 +1,239 @@
+	wg sync.WaitGroup
+	// Messages to the client are routed through the outChan. This ensures
+	// messages are sent in the correct order, and satisfies the thread-safety
+	// requirements of the (*websocket.Conn).WriteMessage.
+	outChan chan []byte
+	// The ban flag will be set if the DEX requests. Upon closing, the client's
+	// IP address will be quarantined by the server if ban = true.
+	ban bool
+}
+
+// newRPCClient is a constructor for a new RPCClient.
+func newRPCClient(addr string, conn wsConnection) *RPCClient {
+	return &RPCClient{
+		on:      true,
+		ip:      addr,
+		conn:    conn,
+		quit:    make(chan struct{}),
+		outChan: make(chan []byte, outBufferSize),
+	}
+}
+
+// SendMessage sends the passed response to the websocket client.  It is backed
+// by a buffered channel, so it will not block until the send channel is full.
+// This approach allows a limit to the number of outstanding requests a client
+// can make without preventing or blocking on async notifications. SendMessage
+// will attempt to marshal the response first, and will return an error if not
+// sucessful.
+func (c *RPCClient) SendMessage(msg interface{}) error {
+	b, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	c.outChan <- b
+	return nil
+}
+
+// start begins processing input and output messages.
+func (c *RPCClient) start() {
+	log.Tracef("Starting websocket client %s", c.ip)
+
+	// Start processing input and output.
+	c.wg.Add(2)
+	go c.inHandler()
+	go c.outHandler()
+}
+
+// disconnect closes the underlying websocket connection.
+func (c *RPCClient) disconnect() {
+	c.quitMtx.Lock()
+	defer c.quitMtx.Unlock()
+	if !c.on {
+		return
+	}
+	c.on = false
+	c.conn.Close()
+	close(c.quit)
+}
+
+// waitForShutdown blocks until the websocket client goroutines are stopped
+// and the connection is closed.
+func (c *RPCClient) waitForShutdown() {
+	c.wg.Wait()
+}
+
+// sendError sends the rpc.RPCError to the client.
+func (c *RPCClient) sendError(id interface{}, jsonErr *rpc.RPCError) error {
+	reply, err := rpc.NewResponse(id, nil, jsonErr)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal reply: %v", err)
+	}
+	err = c.SendMessage(reply)
+	if err != nil {
+		// Something is terribly wrong if rpc.Response is not marshalling.
+		return fmt.Errorf("Failed to marshal Response: %v", err)
+	}
+	return nil
+}
+
+// inHandler handles all incoming messages for the websocket connection. It must
+// be run as a goroutine.
+func (c *RPCClient) inHandler() {
+out:
+	for {
+		// Break out of the loop once the quit channel has been closed.
+		// Use a non-blocking select here so we fall through otherwise.
+		select {
+		case <-c.quit:
+			break out
+		default:
+		}
+		// Block until a message is received or an error occurs.
+		_, msg, err := c.conn.ReadMessage()
+		if err != nil {
+			// Log the error if it's not due to disconnecting.
+			if err != io.EOF {
+				log.Errorf("Websocket receive error from %s: %v", c.ip, err)
+			}
```

Comment:
We'll probably want ping-pong to keep the connection alive.  Otherwise we'd have to recognize the error for read deadline exceeded, and respond by continuing the loop with another read.

Reply:
See how you like https://github.com/decred/dcrdex/pull/33/commits/ee83537d28cbbd97ccd47a1e35dbc413eb909ebb.